syntax = "proto3";

message FontCollectionProto {
    uint32 flavor_proto = 1;
    uint32 header_version_proto = 2;
    map<uint32, FontTableProto> tables_proto = 3;
    repeated FontProto fonts_proto = 4;
}

message GlyphProto {
    int32 x_min_proto = 1;
    int32 x_max_proto = 2;
    int32 y_min_proto = 3;
    int32 y_max_proto = 4;
    uint32 instructions_size_proto = 5;
    bytes instructions_data_proto = 6;
    bool overlap_simple_flag_set_proto = 7;
    repeated ContourVecProto contours_proto = 8;
    bytes composite_data_proto = 9;
    uint32 composite_data_size_proto = 10;
    bool have_instructions_proto = 11;
}

message RebuildMetadataProto {
    uint32 header_checksum_proto = 1;
    repeated WOFF2FontInfoProto font_infos_proto = 2;
    map<uint64, uint32> checksums_proto = 3;
}

message FontProto {
    uint32 flavor_proto = 1;
    uint32 num_tables_proto = 2;
    map<uint32, FontTableProto> tables_proto = 3;
}

message PointProto {
    int32 x_proto = 1;
    int32 y_proto = 2;
    bool on_curve_proto = 3;
}

message TtcFontProto {
    uint32 flavor_proto = 1;
    uint32 dst_offset_proto = 2;
    uint32 header_checksum_proto = 3;
    repeated uint32 table_indices_proto = 4;
}

message BufferProto {
    bytes buffer_proto = 1;
    uint64 length_proto = 2;
    uint64 offset_proto = 3;
}

message WOFF2FontInfoProto {
    uint32 num_glyphs_proto = 1;
    uint32 index_format_proto = 2;
    uint32 num_hmetrics_proto = 3;
    repeated int32 x_mins_proto = 4;
    map<uint32, uint32> table_entry_by_tag_proto = 5;
}

message FontTableProto {
    uint32 tag_proto = 1;
    uint32 checksum_proto = 2;
    uint32 offset_proto = 3;
    uint32 length_proto = 4;
    bytes data_proto = 5;
    repeated uint32 buffer_proto = 6;
    FontTableProto reuse_of_proto = 7;
    uint32 flag_byte_proto = 8;
}

message GlyphPointProto {
    int32 x_proto = 1;
    int32 y_proto = 2;
    bool on_curve_proto = 3;
}

message WOFF2HeaderProto {
    uint32 flavor_proto = 1;
    uint32 header_version_proto = 2;
    uint32 num_tables_proto = 3;
    uint64 compressed_offset_proto = 4;
    uint32 compressed_length_proto = 5;
    uint32 uncompressed_size_proto = 6;
    repeated TableProto tables_proto = 7;
    repeated TtcFontProto ttc_fonts_proto = 8;
}

message TableProto {
    uint32 tag_proto = 1;
    uint32 flags_proto = 2;
    uint32 src_offset_proto = 3;
    uint32 src_length_proto = 4;
    uint32 transform_length_proto = 5;
    uint32 dst_offset_proto = 6;
    uint32 dst_length_proto = 7;
    bytes dst_data_proto = 8;
}

message GlyfEncoderProto {
    repeated uint32 n_contour_stream_proto = 1;
    repeated uint32 n_points_stream_proto = 2;
    repeated uint32 flag_byte_stream_proto = 3;
    repeated uint32 composite_stream_proto = 4;
    repeated uint32 bbox_bitmap_proto = 5;
    repeated uint32 bbox_stream_proto = 6;
    repeated uint32 glyph_stream_proto = 7;
    repeated uint32 instruction_stream_proto = 8;
    repeated uint32 overlap_bitmap_proto = 9;
    int32 n_glyphs_proto = 10;
}

message ContourVecProto {
    repeated GlyphPointProto points_proto = 1;
}
